### PR TITLE
fix(pydantic): Allow conversion of pydantic models with mutable defaults

### PR DESCRIPTION
**Pull request by thejaminator**
## Description

Pydantic allows for defining mutable defaults. This is because behind the scenes, pydantic will call deepcopy / copy on the default value if its mutable.

Now, if we convert this model
```
from pydantic import BaseModel
class User(pydantic.BaseModel):
        friends: List[str] = []

 @strawberry.experimental.pydantic.type(User)
    class UserType:
        friends: strawberry.auto
```
It will error out with

```
ValueError: mutable default <class 'list'> for field friends is not allowed: use default_factory
```

This PR allows such pydantic models to be converted into strawberry. By calling deepcopy / copy on it, and using default_factory instead of default for all defaults.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Allow conversion of pydantic models with mutable defaults
* Fixed issue with not sorting  the defaults that come from default_factory 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: strawberry/experimental/pydantic/object_type.py
Comment (patrick91) [inline strawberry/experimental/pydantic/object_type.py:125]: Can you create a custom exception for this? 😊

File: strawberry/experimental/pydantic/object_type.py
Comment (patrick91) [inline strawberry/experimental/pydantic/object_type.py]: why do we need this change? did we miss some cases?

**Review by patrick91 — COMMENTED** [review]
Looks pretty good to me! I've left a couple of comments :)

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to James Chua for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to James Chua for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/experimental/pydantic/object_type.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/object_type.py:125]: oops, i should probably write a test for this as well

File: strawberry/experimental/pydantic/object_type.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/object_type.py]: yes. i had to add  and model_field[2].default_factory is dataclasses.MISSING because it would only sort out defaults set by a typical default. and it wouldn't sort those defaults added by a default_factory.

i should add a test to prevent such as regression.


also, had to rename field -> model_field because i used field in the early for loop, and mypy was erroring out

Comment (patrick91) [reply strawberry/experimental/pydantic/object_type.py]: the rename sounds good! shall we add a test for these cases too? 😊

Comment (thejaminator) [reply strawberry/experimental/pydantic/object_type.py]: added!

File: strawberry/experimental/pydantic/exceptions.py
Comment (patrick91) [inline strawberry/experimental/pydantic/exceptions.py:30]:
```suggestion
class BothDefaultAndDefaultFactoryDefinedError(Exception):
```
It is a bit long and ugly, but it is better to be explicit 😊

Also we'll do some work in future to improve errors, so we can always rename them later (especially these ones that are not meant to be caught manually)

File: strawberry/experimental/pydantic/utils.py
Comment (patrick91) [inline strawberry/experimental/pydantic/utils.py:9]:
```suggestion
    BothDefaultAndDefaultFactoryDefinedError,
```

File: strawberry/experimental/pydantic/utils.py
Comment (patrick91) [inline strawberry/experimental/pydantic/utils.py:59]: could we use a named tuple for this? 😊

https://docs.python.org/3/library/typing.html#typing.NamedTuple

File: strawberry/experimental/pydantic/utils.py
Comment (patrick91) [inline strawberry/experimental/pydantic/utils.py:80]: nice! thanks for moving these into a different file!

File: tests/experimental/pydantic/test_conversion.py
Comment (patrick91) [inline tests/experimental/pydantic/test_conversion.py:773]: 💯

File: tests/experimental/pydantic/test_conversion.py
Comment (patrick91) [inline tests/experimental/pydantic/test_conversion.py:760]:
```suggestion
        BothDefaultAndDefaultFactoryDefinedError,
```

File: strawberry/experimental/pydantic/object_type.py
Comment (patrick91) [inline strawberry/experimental/pydantic/object_type.py:117]: shall we do something like this to remove one nesting level?

```python
f = (field_name, field for field_name, field in model_fields.items() if field_name in fields_set)
for field_name, field in f:
    # no if here
```

**Review by patrick91 — COMMENTED** [review]
This looks pretty good! 👏 

I've added some additional comments, hope that's ok!

File: strawberry/experimental/pydantic/object_type.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/object_type.py:117]: cool trick to prevent identation hell

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:59]: sure, i wasn't sure which one strawberry would use internally.

**Review by thejaminator — COMMENTED** [review]
add suggest changes

File: strawberry/experimental/pydantic/utils.py
Comment (g-as) [reply strawberry/experimental/pydantic/utils.py:77]: Slightly shorter version
```python
    return sorted(
        fields,
        key=lambda model_field: (
            model_field.field.default is dataclasses.MISSING
            and model_field.field.default_factory is dataclasses.MISSING
        ),
        reverse=True
    )
```

Or with a function to keep annotations.
```python
    def is_missing_default(model_field: DataclassCreationFields):
        """Check if field has defaults."""
        return (
            model_field.field.default is dataclasses.MISSING
            and model_field.field.default_factory is dataclasses.MISSING
        )
    return sorted(fields, key=is_missing_default, reverse=True)
```

Comment (patrick91) [reply strawberry/experimental/pydantic/utils.py:77]: I like the second one, great suggestion 😊

File: strawberry/experimental/pydantic/utils.py
Comment (g-as) [inline strawberry/experimental/pydantic/utils.py:46]: DataclassCreationField~s~ ?

File: strawberry/experimental/pydantic/utils.py
Comment (BryceBeagle) [reply strawberry/experimental/pydantic/utils.py:77]: Another option:
```python
    def has_default(model_field: DataclassCreationFields):
        """Check if field has defaults."""
        return (
            model_field.field.default is not dataclasses.MISSING
            or model_field.field.default_factory is not dataclasses.MISSING
        )
    return sorted(fields, key=has_default)
```

File: strawberry/experimental/pydantic/utils.py
Comment (BryceBeagle) [inline strawberry/experimental/pydantic/utils.py:90]: Looks like this exception needs some coverage

File: strawberry/experimental/pydantic/utils.py
Comment (BryceBeagle) [inline strawberry/experimental/pydantic/utils.py:89]: Should be able to just compare against `arguments.UNSET`

File: strawberry/experimental/pydantic/exceptions.py
Comment (BryceBeagle) [inline strawberry/experimental/pydantic/exceptions.py:31]: Missing coverage

File: tests/experimental/pydantic/test_basic.py
Comment (BryceBeagle) [inline tests/experimental/pydantic/test_basic.py:297]: Why not use `is`/`is not`?

File: tests/experimental/pydantic/test_conversion.py
Comment (BryceBeagle) [inline tests/experimental/pydantic/test_conversion.py:771]: `with` scope should be narrowed to the exact line that raises the exception

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: tests/experimental/pydantic/test_basic.py
Comment (thejaminator) [reply tests/experimental/pydantic/test_basic.py:297]: wouldn't that be "less explicit"? anyone reading the test would need to understand that `is` refers to the reference rather than the hash. most of us would remember that, but for myself i forget the difference from time to time. i'll rather be explicit here so other maintainers can understand it at one glance

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:89]: now the problem is that i can't (i think) refer to the singleton instance of arguments.UNSET in the type signature. so instead my type signature has to refer to the _Unset class. And then since my implementation has to be true to my type signature, then i'll need to check for isinstance

File: tests/experimental/pydantic/test_conversion.py
Comment (thejaminator) [reply tests/experimental/pydantic/test_conversion.py:771]: thanks!

File: tests/experimental/pydantic/test_basic.py
Comment (BryceBeagle) [reply tests/experimental/pydantic/test_basic.py:297]: `a is b` is exactly the same as `id(a) == id(b)`

Comment (thejaminator) [reply tests/experimental/pydantic/test_basic.py:297]: yes correct. i'm not opioniated on this so i'll change it to `is` :D

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:46]: yes. what about it?

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:54]: turns out it needs to be a `Type` object instead

**Comment by thejaminator** [comment]
@BryceBeagle @patrick91 made the changes, please review :). Also, i will be away the next 2 weeks. so if there are any remaining issues / bugs, please feel free to push into my branch. thanks!

File: tests/experimental/pydantic/test_basic.py
Comment (BryceBeagle) [reply tests/experimental/pydantic/test_basic.py]: The first 3 lines should also use `is` as `dataclasses.MISSING` is a singleton object instance:
```
    assert field1.default is dataclasses.MISSING
    assert field2.default is dataclasses.MISSING
    assert field1.default_factory is dataclasses.MISSING
```

Comment (thejaminator) [reply tests/experimental/pydantic/test_basic.py]: changed!

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:88]: seems related to this? https://github.com/strawberry-graphql/strawberry/issues/1473

**Comment by thejaminator** [comment]
> @BryceBeagle @patrick91 made the changes, please review :). Also, i will be away the next 2 weeks. so if there are any remaining issues / bugs, please feel free to push into my branch. thanks!

@BryceBeagle @patrick91 this MR touches a bunch of files. and my workplace wants to move to strawberry but this MR is blocking us. could i have a review soon before it gets stale :D

**Comment by patrick91** [comment]
@thejaminator going to have a look at this in the next few days 😊

File: strawberry/experimental/pydantic/utils.py
Comment (BryceBeagle) [reply strawberry/experimental/pydantic/utils.py:88]:
```patch
Index: strawberry/experimental/pydantic/utils.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/strawberry/experimental/pydantic/utils.py b/strawberry/experimental/pydantic/utils.py
--- a/strawberry/experimental/pydantic/utils.py	(revision 386b1a367a933996d42713067c7621fa01ff2794)
+++ b/strawberry/experimental/pydantic/utils.py	(date 1640309294688)
@@ -4,7 +4,7 @@
 from pydantic.typing import NoArgAnyCallable
 from pydantic.utils import smart_deepcopy
 
-from strawberry.arguments import _Unset
+from strawberry.arguments import UNSET
 from strawberry.experimental.pydantic.exceptions import (
     BothDefaultAndDefaultFactoryDefinedError,
     UnregisteredTypeException,
@@ -84,8 +84,8 @@
     Returns optionally a NoArgAnyCallable representing a default_factory parameter
     """
     final_factory = default_factory
-    if not isinstance(default, _Unset):
-        if not isinstance(default_factory, _Unset):
+    if default is not UNSET:
+        if default_factory is not UNSET:
             raise BothDefaultAndDefaultFactoryDefinedError(
                 default=default, default_factory=default_factory
             )

```
This change passes tests just fine? Are you seeing something else?

File: tests/experimental/pydantic/test_conversion.py
Comment (BryceBeagle) [inline tests/experimental/pydantic/test_conversion.py:774]:
```suggestion
    # should return a factory that copies the default parameter
    assert created_factory() == mutable_default
    assert created_factory() is not mutable_default
```
Nitpick: no reason to over-complicate things here

File: tests/experimental/pydantic/test_conversion.py
Comment (BryceBeagle) [inline tests/experimental/pydantic/test_conversion.py:755]: Couple minor comments here:
1. You added `from arguments import UNSET` at the top of the file, so no reason to grab it from the `arguments` namespace anymore
2. Just add the comments above the line they refer to. The multi-line assert with trailing in-line comments is pretty strange

**Review by BryceBeagle — APPROVED** [review]
Couple minor requests and then this PR is good to go :+1: 

I'll pre-approve the changes so that @patrick91 can merge after he looks at it

File: tests/experimental/pydantic/test_conversion.py
Comment (thejaminator) [reply tests/experimental/pydantic/test_conversion.py]: good catch! resolved :)

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:88]: that works too. I'm just saying that according to `defaults_into_factory` 's type signature, supposedly you could do this

```
my_new_unset = _Unset()
defaults_into_factory(my_new_unset, my_new_unset)
```

However, the function wouldn't work correctly, because my_new_unset is not strawberry's UNSET. 

Anyways, i don't really care about it. _Unset is a private class by the underscore convention anyways.
so i will make the change. 
Guess you guys could have some singleton restriction in the _Unset to get around the typing restriction.

File: strawberry/experimental/pydantic/object_type.py
Comment (patrick91) [inline strawberry/experimental/pydantic/object_type.py:152]: this might change in future https://github.com/strawberry-graphql/strawberry/pull/1187

adding a comment so we can get back to it in the other PR, since it is still open :)

File: strawberry/experimental/pydantic/utils.py
Comment (patrick91) [inline strawberry/experimental/pydantic/utils.py:56]: `to_tuple` sounds like a method :D 

maybe can rename it to `as_tuple` or just `tuple` or remove the property, but can be done in another PR :)

File: RELEASE.md
Comment (patrick91) [inline RELEASE.md:4]: I think there’s a typo here :)
```suggestion
Also fixes bug when converting a pydantic model field with default_factory. Previously it would raise an exception when fields with a default_factory were declared before fields without defaults.
```

**Review by patrick91 — APPROVED** [review]
Looks good! I’ve only added a couple of minor comments, we can merge after fixing them :)

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:55]: removed the property!

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
